### ISSUE TITLE
[Question] In the Jacobian verification for X_dot = J @ q_dot within PhysX, a large error is observed in the linear velocity component.

### ISSUE THREAD

**Issue by littlebearqqq** [isaac-sim/IsaacLab#4521]
### Question

I am currently learning to use Jacobian matrices in Isaac Sim, and I aim to compare the end-effector velocity calculated via the formula Xdot​=J@qdot​ with that obtained directly from Isaac Lab. Yet I cannot figure out why there is a relatively large error in the linear velocity component – this issue occurs specifically for the Unitree G1 humanoid robot I am using, in the scenario based on torso_link. The following are my code and log information:

    def verify_jacobian_left_arm_isaac(self):
        """
        验证左手臂雅可比矩阵是否正确(相对于 base)
        使用公式: \dot{X} = J \dot{q}
        其中:
        - \dot{X}: 末端执行器速度 [线速度(3), 角速度(3)] 在 base 坐标系中、相对于 base
        - J: 雅可比矩阵 [6, n_arm_joints] 在 base 坐标系中
        - \dot{q}: 关节速度 [n_arm_joints]
        这个验证通过 只固定了base 然后腰部关节依然会动 所以要考虑腰部关节 把 self.waist_ids也加上
        """
        robot = self.robot

        # 1. 获取左手臂关节速度
        left_arm_joint_vel = robot.data.joint_vel[:, self.waist_ids + self.left_arm_ids]  # [num_envs, 7]

        # 2. 获取左手末端和 base 的速度 / 姿态（世界坐标系）
        left_wrist_id = self.wrist_body_ids[0]
        left_ee_lin_vel_w = robot.data.body_lin_vel_w[:, left_wrist_id, :]  # [num_envs, 3]
        left_ee_ang_vel_w = robot.data.body_ang_vel_w[:, left_wrist_id, :]  # [num_envs, 3]
        
        left_ee_pos_w = robot.data.body_pos_w[:, left_wrist_id, :]  # [num_envs, 3]
        base_pos_w = robot.data.root_pos_w                          # base在世界系的位置 [num_envs, 3]
        left_ee_pos_rel_w = left_ee_pos_w - base_pos_w              # 末端相对于base的位置（世界系）

        base_lin_vel_w = robot.data.root_lin_vel_w  # [num_envs, 3]
        base_ang_vel_w = robot.data.root_ang_vel_w  # [num_envs, 3]
        base_quat_w = robot.data.root_quat_w        # [num_envs, 4]

        # 3. 计算末端相对于 base 的速度（世界坐标系）
        left_ee_lin_vel_rel_w = left_ee_lin_vel_w - base_lin_vel_w - torch.cross(base_ang_vel_w, left_ee_pos_rel_w)  # [num_envs, 3]
        # left_ee_lin_vel_rel_w = left_ee_lin_vel_w - base_lin_vel_w
        left_ee_ang_vel_rel_w = left_ee_ang_vel_w - base_ang_vel_w  # [num_envs, 3]

        # 4. 转换到 base 坐标系
        left_ee_lin_vel_b = math_utils.quat_apply_inverse(base_quat_w, left_ee_lin_vel_rel_w)  # [num_envs, 3]
        left_ee_ang_vel_b = math_utils.quat_apply_inverse(base_quat_w, left_ee_ang_vel_rel_w)  # [num_envs, 3]
        left_ee_vel_b_actual = torch.cat([left_ee_lin_vel_b, left_ee_ang_vel_b], dim=-1)  # [num_envs, 6]

        # 5. 获取左手臂雅可比矩阵（世界坐标系）
        if self.robot.is_fixed_base:
            left_ee_body_idx = self.wrist_body_ids[0] - 1
        else:
            left_ee_body_idx = self.wrist_body_ids[0]

        mass_matrix_full = robot.root_physx_view.get_generalized_mass_matrices()
        print("mass_matrix_full.shape: ", mass_matrix_full.shape)
        C_v_full = robot.root_physx_view.get_coriolis_and_centrifugal_compensation_forces()
        print("C_v_full.shape: ", C_v_full.shape)
        jacobian_w = robot.root_physx_view.get_jacobians()[:, left_ee_body_idx, :, :]  # [num_envs, 6, n_joints]
        print("fiexd base jacobian_w_full.shape is : ", jacobian_w.shape)
        jacobian_w = jacobian_w[:, :, self.waist_jacobian_ids + self.left_arm_jacobian_ids]  # [num_envs, 6, 7]

        # 6. 将雅可比从世界坐标系转换到 base 坐标系
        base_rot_matrix = matrix_from_quat(quat_inv(base_quat_w))  # [num_envs, 3, 3]
        jacobian_b = jacobian_w.clone()  # [num_envs, 6, 7]
        jacobian_b[:, :3, :] = torch.bmm(base_rot_matrix, jacobian_b[:, :3, :])  # 线速度部分
        jacobian_b[:, 3:, :] = torch.bmm(base_rot_matrix, jacobian_b[:, 3:, :])  # 角速度部分

        # 7. 计算 J @ q_dot
        joint_vel_exp = left_arm_joint_vel.unsqueeze(-1)  # [num_envs, 7, 1]
        left_ee_vel_b_computed = torch.matmul(jacobian_b, joint_vel_exp).squeeze(-1)  # [num_envs, 6]

        # 8. 计算误差
        error = left_ee_vel_b_computed - left_ee_vel_b_actual  # [num_envs, 6]

        # 9. 打印验证结果（仅第一个环境）
        print("isaac左手臂雅可比验证 (环境 0, base坐标系):")
        print(f"isaac计算值 J@q_dot: {left_ee_vel_b_computed[0]}")
        print(f"isaac实际值 X_dot:   {left_ee_vel_b_actual[0]}")
        print(f"isaac误差:           {error[0]}")

        return error
 
    def verify_jacobian_left_arm_isaac_Totorso(self):
        """
        验证左手臂雅可比矩阵是否正确(相对于 torso)
        使用公式: \dot{X} = J \dot{q}
        其中:
        - \dot{X}: 末端执行器速度 [线速度(3), 角速度(3)] 在 torso 坐标系中、相对于 torso
        - J: 雅可比矩阵 [6, n_arm_joints] 在 torso 坐标系中
        - \dot{q}: 关节速度 [n_arm_joints]
        这个验证使用 torso 作为基座，只考虑手臂关节（不包括腰部关节）
        """
        robot = self.robot

        # 1. 获取左手臂关节速度（只包含手臂关节，不包含腰部）
        left_arm_joint_vel = robot.data.joint_vel[:, self.left_arm_ids]  # [num_envs, 7]

        # 2. 获取左手末端和 torso 的速度 / 姿态（世界坐标系）
        left_wrist_id = self.wrist_body_ids[0]
        left_ee_lin_vel_w = robot.data.body_lin_vel_w[:, left_wrist_id, :]  # [num_envs, 3]
        left_ee_ang_vel_w = robot.data.body_ang_vel_w[:, left_wrist_id, :]  # [num_envs, 3]
        
        left_ee_pos_w = robot.data.body_pos_w[:, left_wrist_id, :]  # [num_envs, 3]
        torso_pos_w = robot.data.body_pos_w[:, self.torso_body_id[0], :]  # torso在世界系的位置 [num_envs, 3]
        left_ee_pos_rel_w = left_ee_pos_w - torso_pos_w  # 末端相对于torso的位置（世界系）

        torso_lin_vel_w = robot.data.body_lin_vel_w[:, self.torso_body_id[0], :]  # [num_envs, 3]
        torso_ang_vel_w = robot.data.body_ang_vel_w[:, self.torso_body_id[0], :]  # [num_envs, 3]
        torso_quat_w = robot.data.body_quat_w[:, self.torso_body_id[0], :]  # [num_envs, 4]

        # 3. 计算末端相对于 torso 的速度（世界坐标系）
        left_ee_lin_vel_rel_w = left_ee_lin_vel_w - torso_lin_vel_w - torch.cross(torso_ang_vel_w, left_ee_pos_rel_w)  # [num_envs, 3]
        left_ee_ang_vel_rel_w = left_ee_ang_vel_w - torso_ang_vel_w  # [num_envs, 3]

        # 4. 转换到 torso 坐标系
        left_ee_lin_vel_t = math_utils.quat_apply_inverse(torso_quat_w, left_ee_lin_vel_rel_w)  # [num_envs, 3]
        left_ee_ang_vel_t = math_utils.quat_apply_inverse(torso_quat_w, left_ee_ang_vel_rel_w)  # [num_envs, 3]
        left_ee_vel_t_actual = torch.cat([left_ee_lin_vel_t, left_ee_ang_vel_t], dim=-1)  # [num_envs, 6]

        # 5. 获取左手臂雅可比矩阵（世界坐标系）
        if self.robot.is_fixed_base:
            left_ee_body_idx = self.wrist_body_ids[0] - 1
        else:
            left_ee_body_idx = self.wrist_body_ids[0]

        mass_matrix_full = robot.root_physx_view.get_generalized_mass_matrices()
        #print("mass_matrix_full.shape: ", mass_matrix_full.shape)
        C_v_full = robot.root_physx_view.get_coriolis_and_centrifugal_compensation_forces()
        #print("C_v_full.shape: ", C_v_full.shape)
        jacobian_w = robot.root_physx_view.get_jacobians()[:, left_ee_body_idx, :, :]  # [num_envs, 6, n_joints]
        #print("jacobian_w_full.shape is : ", jacobian_w.shape)
        # 只提取手臂关节对应的雅可比列（不包含腰部关节）
        jacobian_w = jacobian_w[:, :, self.left_arm_jacobian_ids]  # [num_envs, 6, 7]

        # 6. 将雅可比从世界坐标系转换到 torso 坐标系
        torso_rot_matrix = matrix_from_quat(quat_inv(torso_quat_w))  # [num_envs, 3, 3]
        jacobian_t = jacobian_w.clone()  # [num_envs, 6, 7]
        jacobian_t[:, :3, :] = torch.bmm(torso_rot_matrix, jacobian_t[:, :3, :])  # 线速度部分
        jacobian_t[:, 3:, :] = torch.bmm(torso_rot_matrix, jacobian_t[:, 3:, :])  # 角速度部分

        # 7. 计算 J @ q_dot
        joint_vel_exp = left_arm_joint_vel.unsqueeze(-1)  # [num_envs, 7, 1]
        left_ee_vel_t_computed = torch.matmul(jacobian_t, joint_vel_exp).squeeze(-1)  # [num_envs, 6]

        # 8. 计算误差
        error = left_ee_vel_t_computed - left_ee_vel_t_actual  # [num_envs, 6]

        # 9. 打印验证结果（仅第一个环境）
        print("isaac左手臂雅可比验证 (环境 0, torso坐标系):")
        print(f"isaac计算值 J@q_dot: {left_ee_vel_t_computed[0]}")
        print(f"isaac实际值 X_dot:   {left_ee_vel_t_actual[0]}")
        print(f"isaac误差:           {error[0]}")

        return error

log:
```
    mass_matrix_full.shape:  torch.Size([4096, 35, 35])
    C_v_full.shape:  torch.Size([4096, 35])
    fiexd base jacobian_w_full.shape is :  torch.Size([4096, 6, 35])
    isaac左手臂雅可比验证 (环境 0, base坐标系):
    isaac计算值 J@q_dot: tensor([-0.9003, -0.8812, -0.8439, -1.3323,  1.0512, -0.2642], device='cuda:0')
    isaac实际值 X_dot:   tensor([-0.9157, -0.8025, -0.8724, -1.3323,  1.0512, -0.2642], device='cuda:0')
    isaac误差:           tensor([ 1.5473e-02, -7.8671e-02,  2.8523e-02,  3.5763e-07,  3.5763e-07,
             4.4703e-07], device='cuda:0')
    isaac左手臂雅可比验证 (环境 0, torso坐标系):
    isaac计算值 J@q_dot: tensor([-0.5884, -1.1627, -0.6537, -1.6483,  2.0859, -1.9222], device='cuda:0')
    isaac实际值 X_dot:   tensor([-0.4464, -0.9480, -0.6174, -1.6483,  2.0859, -1.9222], device='cuda:0')
    isaac误差:           tensor([-1.4197e-01, -2.1477e-01, -3.6293e-02,  1.1921e-07,  0.0000e+00,
             5.9605e-07], device='cuda:0')
```

- Isaac Lab Version: [e.g. 2.3.2]
- Isaac Sim Version: [e.g. 5.1, this can be obtained by `cat ${ISAACSIM_PATH}/VERSION`]

**Comment by RandomOakForest** [comment]
Thank you for posting this. Let us know if you need further help.
